src/libcmis/oauth2-providers.cxx

```diff
@@ -292,18 +299,22 @@ int OAuth2Providers::parseResponse ( const char* response, string& post, string&
             xmlChar* action = xmlTextReaderGetAttribute( reader, 
                                                          BAD_CAST( "action" ));
 
-            // GDrive pin code page contains 3 forms.
-            // First form resends pin code, we need to omit its input fields.
-            // Also we mustn't parse action field from the 'select challenge' form.
+            // GDrive pin code page contains many forms.
+            // We have to parse only the form with pin field.
             if ( action != NULL )
             {
-                if ( strcmp((char*)action, "/signin/challenge") != 0
-                   && strcmp((char*)action, "/signin/selectchallenge") != 0
-                   && xmlStrlen(action) > 0 )
+                bool bChallengePage = ( xmlStrlen( action ) > 0 )
+                        && ( strncmp( (char*)action, "/signin", strlen( "/signin" ) ) == 0 );
```

Comment:
Ah, right, I missed this. It makes sense then. But it would be better to define a macro that expands to the string literal and its length to avoid repeating the literal. And to use sizeof instead of strlen for computing the length.